runtimebp/cpu.go

```diff
@@ -48,52 +39,17 @@ func NumCPU() (n float64) {
 		return
 	}
 
-	period, err = readNumberFromFile(periodPath, buf)
-	if err != nil {
+	// CFS quota returns -1 if there is no limit, return the default.
+	if quota < 0 {
 		return
 	}
 
-	return quota / period
-}
-
-// On some situations the quota file will be -1,
-// and we should use this one instead.
-//
-// Those situations include:
-//
-// - Very old version of docker
-//
-// - In k8s only request is set for cpu, not limit
-func numCPUSharesFallback() (n float64) {
-	const (
-		sharesPath  = "/sys/fs/cgroup/cpu/cpu.shares"
-		denominator = 1024
-	)
-
-	var err error
-	defer func() {
-		if err != nil || n <= 0 {
-			// Fallback and log to stderr.
-			fmt.Fprintf(
-				os.Stderr,
-				"NumCPU: falling back to use runtime.NumCPU(): %v, %v\n",
-				n,
-				err,
-			)
-			n = float64(runtime.NumCPU())
-		}
-	}()
-
-	// Big enough buffer to read the number in the file wholly into memory.
-	buf := make([]byte, 1024)
-	var shares float64
-
-	shares, err = readNumberFromFile(sharesPath, buf)
+	period, err = readNumberFromFile(periodPath, buf)
 	if err != nil {
 		return
 	}
 
-	return float64(shares) / denominator
+	return quota / period
 }
```

Comment:
💭 While we're refactoring this, worth making it testable as well? Could use an [`fs.FS`](https://pkg.go.dev/io/fs#FS) + pull the `runtime.NumCPU` dependency up into a `func() int` variable that could be overwritten in tests.